karapace/kafka_rest_apis/__init__.py

```diff
@@ -739,18 +742,82 @@ def is_valid_schema_request(data: dict, prefix: str) -> bool:
                 return False
         return isinstance(schema, str)
 
-    async def get_schema_id(self, data: dict, topic: str, prefix: str, schema_type: SchemaType) -> int:
-        log.debug("Retrieving schema id for %r", data)
-        if f"{prefix}_schema_id" in data and data[f"{prefix}_schema_id"] is not None:
-            log.debug("Will use schema id %d for serializing %s on topic %s", data[f"{prefix}_schema_id"], prefix, topic)
-            return int(data[f"{prefix}_schema_id"])
-        schema_str = data[f"{prefix}_schema"]
-        log.debug("Registering / Retrieving ID for schema %s", schema_str)
-        if schema_str not in self.schemas_cache:
-            subject_name = self.serializer.get_subject_name(topic, data[f"{prefix}_schema"], prefix, schema_type)
-            schema_id = await self.serializer.get_id_for_schema(data[f"{prefix}_schema"], subject_name, schema_type)
-            self.schemas_cache[schema_str] = schema_id
-        return self.schemas_cache[schema_str]
+    async def get_schema_id(
+        self,
+        data: dict,
+        topic: str,
+        prefix: str,
+        schema_type: SchemaType,
+    ) -> SchemaId:
+        """
+        This method search and validate the SchemaId for a request, it acts as a guard (In case of something wrong
+        throws an error).
+
+        :raises InvalidSchema:
+        """
+        log.debug("[resolve schema id] Retrieving schema id for %r", data)
+        schema_id: Union[SchemaId, None] = (
+            SchemaId(int(data[f"{prefix}_schema_id"])) if f"{prefix}_schema_id" in data else None
+        )
+        schema_str = data.get(f"{prefix}_schema")
+
+        if schema_id is None and schema_str is None:
+            raise InvalidSchema()
+
+        if schema_id is None:
+            parsed_schema = ValidatedTypedSchema.parse(schema_type, schema_str)
+            subject_name = self.serializer.get_subject_name(topic, parsed_schema, prefix, schema_type)
+            schema_id = await self._query_schema_id_from_cache_or_registry(parsed_schema, schema_str, subject_name)
+        else:
+            parsed_schema, valid_subjects = await self._query_schema_and_subjects(schema_id, skip_cache=False)
+            subject_name = self.serializer.get_subject_name(topic, parsed_schema, prefix, schema_type)
+
+            if subject_name not in valid_subjects:
+                # first checking again the result by contacting directly the schema registry
+                _, updated_subjects = await self._query_schema_and_subjects(schema_id, skip_cache=True)
+                if subject_name not in updated_subjects:
+                    raise InvalidSchema()
```

Comment:
thought: It looks like this strategy means that the worst case will produce two over-the-network calls. If we instead encapsulated this logic inside the serializer, it seems like we should be able to determine whether the first try used the cache or not, and that way eliminate the need for one extra call?

Reply:
not so trivially since we need the typed schema to compute the subject, but I found a way to always do only 1 network call :) if we pass down a function that having the schema and the list of allowed subjects can tell if the subject it's present or not we can always do the call if the result isn't cached or we can do the call only if the subject isn't present in the allowed subjects from the cache.